src/docker/remote.rs

```diff
@@ -0,0 +1,653 @@
+    }
+
+    // now we need to copy over the host toolchain too, since it has
+    // some requirements to find std libraries, etc.
+    let rustc = sysroot.join("bin").join("rustc");
+    let libdir = Command::new(rustc)
+        .args(&["--print", "target-libdir"])
+        .run_and_get_stdout(verbose)?;
+    let host_toolchain_path = Path::new(libdir.trim()).parent().unwrap();
+    copy_volume_files(
+        engine,
+        container,
+        host_toolchain_path,
+        &dst_rustlib,
+        verbose,
+    )?;
+
+    Ok(())
+}
+
+pub fn volume_create(engine: &Engine, volume: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "volume")
+        .args(&["create", volume])
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn volume_rm(engine: &Engine, volume: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "volume")
+        .args(&["rm", volume])
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn volume_exists(engine: &Engine, volume: &str, verbose: bool) -> Result<bool> {
+    subcommand(engine, "volume")
+        .args(&["inspect", volume])
+        .run_and_get_output(verbose)
+        .map(|output| output.status.success())
+        .map_err(Into::into)
+}
+
+pub fn container_stop(engine: &Engine, container: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "stop")
+        .arg(container)
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn container_rm(engine: &Engine, container: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "rm")
+        .arg(container)
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn container_state(engine: &Engine, container: &str, verbose: bool) -> Result<ContainerState> {
+    let stdout = subcommand(engine, "ps")
+        .arg("-a")
+        .args(&["--filter", &format!("name={container}")])
+        .args(&["--format", "{{.State}}"])
+        .run_and_get_stdout(verbose)?;
+    ContainerState::new(stdout.trim())
+}
+
+fn path_hash(path: &Path) -> String {
+    sha1_smol::Sha1::from(path.display().to_string().as_bytes())
+        .digest()
+        .to_string()
+        .get(..5)
+        .expect("sha1 is expected to be at least 5 characters long")
+        .to_string()
+}
+
+pub fn unique_container_identifier(
+    target: &Target,
+    metadata: &CargoMetadata,
+    dirs: &Directories,
+) -> Result<String> {
+    let host_version_meta = rustc::version_meta()?;
+    let commit_hash = host_version_meta
+        .commit_hash
+        .unwrap_or(host_version_meta.short_version_string);
+
+    let workspace_root = &metadata.workspace_root;
+    let package = metadata
+        .packages
+        .iter()
+        .find(|p| p.manifest_path.parent().unwrap() == workspace_root)
+        .unwrap_or_else(|| metadata.packages.get(0).unwrap());
+
+    let name = &package.name;
```

Comment:
is this unwrap safe?

Also, can we apply this logic in `Dockerfile::image_name` :)

Reply:
if it is safe, use `expect` instead